python/phase-2.md

```diff
@@ -0,0 +1,831 @@
+
+@workflow.defn
+class GreetingWorkflow:
+    def __init__() -> None:
+        self._salutation = "Hello"
+        self._salutation_update = asyncio.Event()
+        self._complete = asyncio.Event()
+
+    @workflow.run
+    async def run(self, name: str) -> str:
+        while True:
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
+            
+            # Wait for salutation update or complete signal (this can be
+            # cancelled)
+            await asyncio.wait(
+                [self._salutation_update.wait(), self._complete.wait()],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if self._salutation_update.is_set():
+                self._salutation_update.clear()
+            if self._complete.is_set():
+                return self._current_greeting
+
+    @workflow.signal
+    async def update_salutation(self, salutation: str) -> None:
+        self._salutation = salutation
+        self._salutation_update.set()
+
+    @workflow.signal
+    async def complete_with_greeting(self) -> None:
+        self._complete.set()
+
+    @workflow.query
+    async def current_greeting(self) -> str:
+        return self._current_greeting
+
+def test_workflow(client: Client) -> None:
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
+
+    # Change the salutation part
+    await handle.signal(GreetingWorkflow.update_salutation, "Hola")
+
+    # Now tell it to complete with the new greeting
+    await handle.signal(GreetingWorkflow.complete_with_greeting)
+
+    # Check result
+    assert "Hola, John!" == await handle.result()
+
+def create_worker(client: Client) -> Worker:
+    return Worker(
+      client=client,
+      activities=[create_greeting_activity],
+      workflows=[GreetingWorkflow],
+    )
+```
+
+Notes:
+
+* Workflows are classes
+  * Consideration was given to making workflows be functions (in addition to classes or having the class equivalent sit
+    on top) with several possibilities for defining signals and queries, but all were too ugly/complicated
+* `@activity.defn`
+  * Must be on function and accepts an optional activity `name` which defaults to function `__name__`
+  * This was originally not wanted in the activity impl (phase 1) due to complaints of the name not being obvious
+  * We need to define the activity name on the activity for execution by the workflow
+  * Is it ok to now make this required? Is it ok to now make the name implicit? Is it ok to now use a list instead of a
```

Comment:
I'm not a fan of this decorator.
It'd be much more practical to have some collection class where you could accumulate these grouped activities:

```python
coll = ActivityCollection()

@coll.activity
def my_activity():
    ...

Worker(activities=coll)
```

The above is a common pattern in Python frameworks.
That being said, I wouldn't want to limit to only using this collection concept and I'd rather let users decide how they want to register their activities.

If we want to keep this decorator for allowing the name override I'd be fine with that but I wouldn't force this convention.

Reply:
Why do we have to force global state on people? I just need a simple marker.

> I'd rather let users decide how they want to register their activities.

This is the opposite conclusion that was previously reached which was that we don't want a bunch of different ways to register

> If we want to keep this decorator for allowing the name override I'd be fine with that but I wouldn't force this convention.

Why not require the decorator on every activity and on every workflow? What benefit is there to allowing activities be not marked as such?

This middleground approach was chosen because I was repeatedly told we want to be explicit on the name and this is the only way I can think of having the name be somewhat explicit (you know it's named what's beneath the decorator unless set).

Reply:
> I'd rather let users decide how they want to register their activities.

Why? I just don't at all follow why allowing for a bunch of different ways to do something required is a good thing for users. This should be a zero-thought operation. You register your activities this way, do it once, forget about it. Choice just increases the user's cognitive load and I don't follow what the benefit is to allowing bunches of different ways to do it.

Explicit and obvious for required, only-once operations seems to me a very good thing. Set it and forget it.

Reply:
My main point is that we should be flexible enough to support different registration methods.
If users want to use a "collection" the API should support it, I never said we should provide a collection primitive.

If the worker accepts a mapping or an iterable that should be flexible enough.

If y'all want a mandatory decorator on activity functions I'm fine with that although I don't see the value apart from overriding the activity name TBH.

Reply:
>  If users want to use a "collection" the API should support it, I never said we should provide a collection primitive.

I think it's ok to ask them to convert to a list. The user can do this:

```python
my_activities = {"act-1": my_func1, "act-2": my_func2}
to_register = [activity.defn(fn, name=name) for name, fn in my_activities]
```